### PR TITLE
[quant][core][gpu][improvement] Set tensors as virtual in quantized cudnn linear op

### PR DESCRIPTION
**Pull request by dzdang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75969
* __->__ #75967

Summary:
For cudnn, if certain intermediate tensors are not needed for other
parts of the operation graph in cudnn, the it can be set as virtual, so
that a pointer-uid pair does not need to be provided for the variant
pack, allowing for better performance.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35717354](https://our.internmc.facebook.com/intern/diff/D35717354)

### CODE REVIEW

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35717354).

File: aten/src/ATen/native/quantized/cudnn/Linear.cpp [hunk 1]
```diff
@@ -147,26 +147,15 @@ void PackedLinearWeightCudnn::apply_impl_helper(const at::Tensor& quantized_outp
...
-    uids.reserve(10);
-    data_ptrs = {input_fp.data_ptr(), linear_output.data_ptr(),
-                                           weight_fp.data_ptr(),
-                                           requantize_multiplier_tensor.data_ptr(),
-                                           reinterpret_cast<int8_t*>(quantized_output.data_ptr())};
-    uids = {'x', 'y', 'w', 's', 'r'};
+    data_ptrs.reserve(9);
+    uids.reserve(9);
+    data_ptrs = {input_fp.data_ptr(), weight_fp.data_ptr(), requantize_multiplier_tensor.data_ptr(),
+                 reinterpret_cast<int8_t*>(quantized_output.data_ptr())};
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cudnn/Linear.cpp:153]: can we use data_ptr<int8_t> now?

File: aten/src/ATen/native/quantized/cudnn/Linear.cpp [hunk 2]
```diff
@@ -237,7 +228,7 @@ void PackedLinearWeightCudnn::apply_impl_helper(const at::Tensor& quantized_outp
     // we use inplace operation here where the output is assigned to the input
     relu_op.emplace(cudnn_frontend::OperationBuilder(CUDNN_BACKEND_OPERATION_POINTWISE_DESCRIPTOR)
       .setxDesc(tensor2requant_ptr)
-      .setyDesc(cudnn_utils::getTensorDescriptor(linear_output, 'f', key.output_alignment))
+      .setyDesc(cudnn_utils::getTensorDescriptor(linear_output, 'f', key.output_alignment, true))
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cudnn/Linear.cpp:231]: is this the final output? so setting final output as virtual would speed things up?

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35717354).

File: aten/src/ATen/native/quantized/cudnn/utils.h [hunk 3]
```diff
@@ -193,9 +193,19 @@ uint8_t getAlignment(const at::Tensor &t) {
   return alignment;
 }
 
-cudnn_frontend::Tensor getTensorDescriptor(const at::Tensor &t, int64_t id, uint8_t alignment) {
+cudnn_frontend::Tensor getTensorDescriptor(const at::Tensor &t, int64_t id, uint8_t alignment, bool is_virtual = false) {
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cudnn/utils.h:196]: please add some docs for `is_virtual` here

**Review by jerryzh168 — APPROVED** [review]

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35717354).

**Pull request closed without merging**